### PR TITLE
Update build scripts in Templates to leverage the Cake.MinVer add-in for Cake

### PR DESCRIPTION
**Pull request by augustoproiete**
@RehanSaeed As per our conversation over email. Took me a little while to get to this one but here it is :smile:

### CODE REVIEW

File: Source/ApiTemplate/build.cake [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#addin "nuget:?package=Cake.MinVer&version=1.0.0-rc0001"
```
Comment (RehanSaeed) [inline Source/ApiTemplate/build.cake:1]: This needs to be:

```c#
//#if (Docker)
 #addin "nuget:?package=Cake.MinVer&version=1.0.0-rc0001" 

//#endif
```

**Comment by augustoproiete** [comment]
ps: I see a couple of builds failed but doesn't seem to be related to the contents of the PR.

**Review by RehanSaeed — COMMENTED** [review]
Honestly, I'm still in two minds about accepting this.

- On the one hand it does make the code simpler :).
- On the other hand I've used the `Cake.Docker` plugin before and had my CI build fail because a new version of Cake made breaking changes which broke the plugin. It took quite some time for a new update to be released that was compatible with the latest version of Cake.
- We could use https://github.com/adamralph/simple-exec which would have a very similar PR but maybe look slightly more complex but it wouldn't break the CI build when a new version of Cake has breaking changes.

@VictorioBerra What do you think? Leave as is, use simple-exec or use this plugin?

You can ignore the PR build failure. Orleans builds fail on Mac sometimes. I'm trying to get to the bottom of it.

File: Source/ApiTemplate/.gitignore [hunk 2]
```diff
@@ -308,9 +308,9 @@ paket-files/
 __pycache__/
 *.pyc
 
-# Cake - Uncomment if you are using it
-# tools/**
-# !tools/packages.config
+# Cake
```
[the remark this answers is not in the document]
Comment (augustoproiete) [reply Source/ApiTemplate/.gitignore:311]: The Cake.MinVer plugin gets downloaded to the `tools` folder during build so you do want to ignore the `tools` folder for sure!

File: Source/ApiTemplate/build.cake (hunk 1 above)
Comment (augustoproiete) [reply Source/ApiTemplate/build.cake:1]: Does that apply to all 3 scripts? I'll wait for you to decide if you'll move forward with it

**Comment by augustoproiete** [comment]
@RehanSaeed I'm obviously biased of course :upside_down_face:, but I would argue that given that you decided to use Cake to orchestrate the build of the templates, leveraging Cake plugins is the logical next step and one of the things that makes Cake such a great tool, and that the plugin definitely makes the build script simpler.

I hear you re: breaking changes, and as with any dependency that can happen. I use Cake.MinVer myself on several projects so you can expect it to be kept up-to-date with new Cake versions.

The Cake website now shows you which Cake versions an add-in is compatible with, so that is something you can check before updating Cake, to make sure any plugins you use are compatible with the new version (and, of course, update the plugins to the latest version as well).

[image](https://cakebuild.net/extensions/cake-minver/)

https://cakebuild.net/extensions/cake-minver/

Lastly, the Cake plugin now belongs to the Cake Contrib organization and the core team can make releases, if I get hit by a bus or something :smiley: so there's a good chance it won't be abandoned.

**Comment by VictorioBerra** [comment]
@RehanSaeed @augustoproiete 

I like it. +1 for Cake MinVer Plugin.

- > Lastly, the Cake plugin now belongs to the Cake Contrib organization and the core team can make new releases, if I get hit by a bus or something 😃 so there's a good chance it won't be abandoned.
  - This is really good to know, and will help a lot with cake versions breaking. If they/you ship the CLI tool and the Cake CLI, and the Cake MinVer plugin all at the same time, it will be pretty unlikely we see breaking changes.
 - > I've used the Cake.Docker plugin before and had my CI build fail because a new version of Cake made breaking changes which broke the plugin.
   - Luckily this is super easy to fix and its all self contained within the cake script, we could just throw in SimpleExec and be done.
 - > leveraging Cake plugins is the logical next step and one of the things that makes Cake such a great tool
   - Totally agree. Cake will resolve the nuget package at runtime which replaces `dotnet tool restore`.
   
Currently we can run MinVer without invoking cake, but cake allows running a specific task so this should not be an issue right? Do we even care to do that?

**Comment by augustoproiete** [comment]
Cheers @VictorioBerra

> Currently we can run MinVer without invoking cake, but cake allows running a specific task so this should not be an issue right?

[Cake.MinVer](https://github.com/cake-contrib/Cake.MinVer) effectively runs the [minver-cli](https://www.nuget.org/packages/minver-cli/)) from anywhere it can find (Cake tools path, dotnet local tool, dotnet global tool, custom path, ...) and parses its content to make it available to the Cake build, so you don't "need" to abandon your current approach of installing the local too + `dotnet tool restore` in order to use Cake.MinVer - unless you want to.

This PR assumes that you want to continue with using minver as a local tool + `dotnet tool restore`, and only replaces the execution and parsing of MinVer from within the Cake scripts that do that, to leverage Cake.MinVer.

If you do decide to go the route of using Cake to install the minver-cli nuget package instead, then yes you could create a target inside the Cake script that you can run to execute MinVer exclusively.

**Comment by VictorioBerra** [comment]
I see, I like leaving the process execution, argument passing, output wrangling to a cake addon vs baking it into our cake file. Even though we have a minimal approach.

File: Source/ApiTemplate/.gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (RehanSaeed) [reply Source/ApiTemplate/.gitignore:311]: I didn't realize that Cake does this with tools. I thought Cake.Tool had the advantage of not creating that folder.

I quite often have a Tools directory in my solutions and it's very annoying that Cake pollutes your directory structure like this. At the very least they should create a hidden `.cake` folder like Visual Studio creates `.vs`. You can rename the folder but it's not very clean.

Just read https://github.com/cake-build/cake/discussions/2979. @augustoproiete looks like you got there first! I think we should wait for that to go in before we complete this PR.

[the remark this answers is not in the document]
Comment (augustoproiete) [reply Source/ApiTemplate/.gitignore:311]: Cheers @RehanSaeed.

I'll leave it to you. It's fairly easy to include a `cake.config` with DotNet-Boxed that changes the folder to `.cake`. You can see an [example here](https://github.com/cake-contrib/Cake.MinVer/blob/0b97527398ad36a0a191ab7514280e0f34090efe/cake.config#L7-L9) (and the updated [.gitignore here](https://github.com/cake-contrib/Cake.MinVer/blob/0b97527398ad36a0a191ab7514280e0f34090efe/.gitignore#L26-L27)). I could update this PR with that if make sense.

**Pull request closed without merging**